rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,418 @@
+counter can be incremented easily.
+
+As an example, we can consider provisioning several bots. We'll need to account
+for future overrides so we can fix a single bot in the future if needed:
+
+``` hcl
+locals {
+  nodes = toset(["foo", "bar", "baz"])
+}
+
+resource "teleport_bot" "example" {
+  name = "example"
+  roles = ["access"]
+}
+
+variable "bot_rejoin_overrides" {
+  type = map(number)
+  default = {
+    foo = 5
+  }
+}
+
+resource "teleport_provision_token" "example" {
+  for_each = local.nodes
+
+  version = "v2"
+  metadata = {
+    name = "example-${each.key}"
+  }
+  spec = {
+    roles = ["Bot"]
+    bot_name = teleport_bot.example.name
+    join_method = "challenge"
+
+    challenge = {
+      rejoining = {
+        # look up node-specific count in the rejoin overrides map, default to 2
+        total_rejoins = lookup(var.bot_rejoin_overrides, each.key, 2)
+      }
+    }
+  }
+}
+
+resource "aws_instance" "example" {
+  for_each = teleport_provision_token.example
+
+  ami           = "ami-12345678"
+  instance_type = "t2.micro"
+
+  user_data = templatefile("${path.module}/user-data.tpl", {
+    initial_join_token = each.value.status.challenge.initial_join_secret
+  })
+}
+```
+
+In this example, if node `bar` uses its 2 renewals, we can add a new entry for
+it in `bot_rejoin_overrides` and its `ProvisionToken` will be updated to allow
+additional renewals.
+
+#### Challenge Ceremony
+
+The challenge ceremony will take inspiration from several existing join methods.
+We can create an interactive challenge similar to TPM joining and present a
+challenge containing a nonce. To avoid completely implementing our own
+authentication ceremony, clients can use `go-jose` to marshal and sign a JWT
+which can then be verified easily on the server.
+
+TODO: This needs significant further elaboration and feedback.
+
+#### Client-Side Changes in `tbot`
+
+Bots should be informed of their number of remaining rejoins. We can give bots
+permission to view their own join token, or include the number of remaining
+rejoins as an informational field in the bot's current user certificate. We
+should then expose this as a Prometheus metric to allow for alerting if a bot
+
+We should also investigate hardware key storage backends using HSMs / PKCS#11.
+We have some prior art here in Teleport's [HSM
+support](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/).
+
+#### Non-Terraform UX
+
+TODO
+
+#### Remaining Downsides
+
+- Repairing a bot that has exhausted all of its rejoins is still a semi-manual
+  process. It is significantly easier, and does not necessarily require any
+  changes on the impacted bot node itself, but is still annoying. Users can opt
+  out of this by setting `.spec.rejoining.unlimited=true`, but this has obvious
+  security implications.
+
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
```

Comment:
For organizations that desire multi-use tokens because it alleviates pain for a central administrator who must vend these, I think one supporting improvement would be to allow more tightly scoped rbac on token CRUD based on the role/bot_name.

Reply:
Hmm, this would be fantastic, I'll add a section for this.

Reply:
I've added a section, but I think this probably depends on [Scoped RBAC](https://github.com/gravitational/teleport/pull/38078)